<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Persiapan-Data" data-toc-modified-id="Persiapan-Data-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Persiapan Data</a></span><ul class="toc-item"><li><span><a href="#Library" data-toc-modified-id="Library-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Library</a></span></li><li><span><a href="#Path" data-toc-modified-id="Path-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Path</a></span></li><li><span><a href="#Informasi-Umum-Data" data-toc-modified-id="Informasi-Umum-Data-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Informasi Umum Data</a></span></li><li><span><a href="#Sanity-check-setiap-fitur" data-toc-modified-id="Sanity-check-setiap-fitur-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Sanity check setiap fitur</a></span><ul class="toc-item"><li><span><a href="#Fitur-VehicleType" data-toc-modified-id="Fitur-VehicleType-1.4.1"><span class="toc-item-num">1.4.1&nbsp;&nbsp;</span>Fitur VehicleType</a></span></li><li><span><a href="#Fitur-Gearbox" data-toc-modified-id="Fitur-Gearbox-1.4.2"><span class="toc-item-num">1.4.2&nbsp;&nbsp;</span>Fitur Gearbox</a></span></li><li><span><a href="#Fitur-Power" data-toc-modified-id="Fitur-Power-1.4.3"><span class="toc-item-num">1.4.3&nbsp;&nbsp;</span>Fitur Power</a></span></li><li><span><a href="#Fitur-Model" data-toc-modified-id="Fitur-Model-1.4.4"><span class="toc-item-num">1.4.4&nbsp;&nbsp;</span>Fitur Model</a></span></li><li><span><a href="#Fitur-RegistrationYear" data-toc-modified-id="Fitur-RegistrationYear-1.4.5"><span class="toc-item-num">1.4.5&nbsp;&nbsp;</span>Fitur RegistrationYear</a></span></li><li><span><a href="#Fitur-Mileage" data-toc-modified-id="Fitur-Mileage-1.4.6"><span class="toc-item-num">1.4.6&nbsp;&nbsp;</span>Fitur Mileage</a></span></li><li><span><a href="#Fitur-RegistrationMonth" data-toc-modified-id="Fitur-RegistrationMonth-1.4.7"><span class="toc-item-num">1.4.7&nbsp;&nbsp;</span>Fitur RegistrationMonth</a></span></li><li><span><a href="#Fitur-FuelType" data-toc-modified-id="Fitur-FuelType-1.4.8"><span class="toc-item-num">1.4.8&nbsp;&nbsp;</span>Fitur FuelType</a></span></li><li><span><a href="#Fitur-Brand" data-toc-modified-id="Fitur-Brand-1.4.9"><span class="toc-item-num">1.4.9&nbsp;&nbsp;</span>Fitur Brand</a></span></li><li><span><a href="#Fitur-NotRepaired" data-toc-modified-id="Fitur-NotRepaired-1.4.10"><span class="toc-item-num">1.4.10&nbsp;&nbsp;</span>Fitur NotRepaired</a></span></li><li><span><a href="#Fitur-NumberOfPictures" data-toc-modified-id="Fitur-NumberOfPictures-1.4.11"><span class="toc-item-num">1.4.11&nbsp;&nbsp;</span>Fitur NumberOfPictures</a></span></li></ul></li><li><span><a href="#Perbaikan-Data" data-toc-modified-id="Perbaikan-Data-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Perbaikan Data</a></span><ul class="toc-item"><li><span><a href="#Drop" data-toc-modified-id="Drop-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Drop</a></span></li><li><span><a href="#missing-value" data-toc-modified-id="missing-value-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>missing value</a></span></li><li><span><a href="#Pemisahan-data" data-toc-modified-id="Pemisahan-data-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Pemisahan data</a></span></li></ul></li></ul></li><li><span><a href="#Pelatihan-Model" data-toc-modified-id="Pelatihan-Model-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Pelatihan Model</a></span><ul class="toc-item"><li><span><a href="#Train-test-split-data" data-toc-modified-id="Train-test-split-data-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Train test split data</a></span><ul class="toc-item"><li><span><a href="#Data-OHE" data-toc-modified-id="Data-OHE-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Data OHE</a></span></li><li><span><a href="#Data-Category" data-toc-modified-id="Data-Category-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Data Category</a></span></li></ul></li><li><span><a href="#Sanity-Check" data-toc-modified-id="Sanity-Check-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Sanity Check</a></span></li><li><span><a href="#Fungsi-check-hasil-RMSE" data-toc-modified-id="Fungsi-check-hasil-RMSE-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Fungsi check hasil RMSE</a></span></li><li><span><a href="#DecisionTreeRegressor" data-toc-modified-id="DecisionTreeRegressor-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>DecisionTreeRegressor</a></span></li><li><span><a href="#RandomForestRegressor" data-toc-modified-id="RandomForestRegressor-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>RandomForestRegressor</a></span></li><li><span><a href="#LightGBM" data-toc-modified-id="LightGBM-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>LightGBM</a></span><ul class="toc-item"><li><span><a href="#Data-OHE" data-toc-modified-id="Data-OHE-2.6.1"><span class="toc-item-num">2.6.1&nbsp;&nbsp;</span>Data OHE</a></span></li><li><span><a href="#Data-Category" data-toc-modified-id="Data-Category-2.6.2"><span class="toc-item-num">2.6.2&nbsp;&nbsp;</span>Data Category</a></span></li></ul></li><li><span><a href="#CatBoost" data-toc-modified-id="CatBoost-2.7"><span class="toc-item-num">2.7&nbsp;&nbsp;</span>CatBoost</a></span><ul class="toc-item"><li><span><a href="#Data-OHE" data-toc-modified-id="Data-OHE-2.7.1"><span class="toc-item-num">2.7.1&nbsp;&nbsp;</span>Data OHE</a></span></li><li><span><a href="#Data-Category" data-toc-modified-id="Data-Category-2.7.2"><span class="toc-item-num">2.7.2&nbsp;&nbsp;</span>Data Category</a></span></li></ul></li></ul></li><li><span><a href="#Analisis-Model" data-toc-modified-id="Analisis-Model-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Analisis Model</a></span><ul class="toc-item"><li><span><a href="#Sanity-Check-Linear-Regression" data-toc-modified-id="Sanity-Check-Linear-Regression-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Sanity Check Linear Regression</a></span></li><li><span><a href="#DecisionTreeRegressor" data-toc-modified-id="DecisionTreeRegressor-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>DecisionTreeRegressor</a></span></li><li><span><a href="#RandomForestRegressor" data-toc-modified-id="RandomForestRegressor-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>RandomForestRegressor</a></span></li><li><span><a href="#LightGBM" data-toc-modified-id="LightGBM-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>LightGBM</a></span></li><li><span><a href="#CatBoost" data-toc-modified-id="CatBoost-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>CatBoost</a></span></li><li><span><a href="#Kesimpulan" data-toc-modified-id="Kesimpulan-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Kesimpulan</a></span></li></ul></li></ul></div>

Rusty Bargain adalah perusahaan jual beli mobil bekas yang sedang mengembangkan sebuah aplikasi untuk menarik pembeli baru. Dalam aplikasi tersebut, Anda bisa dengan cepat mengetahui nilai pasar mobil Anda. Anda memiliki akses ke data historis, spesifikasi teknis kendaraan, versi model kendaraan, dan harga kendaraan. Tugas Anda adalah membuat suatu model yang mampu menentukan nilai pasar mobil.

Rusty Bargain tertarik pada:

- kualitas prediksi;
- kecepatan model dalam memprediksi;
- waktu yang diperlukan untuk melatih model

## Persiapan Data

### Library

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
import lightgbm as lgb
from catboost import CatBoostRegressor, Pool

### Path

In [ ]:
path = '/datasets/car_data.csv'

In [ ]:
df = pd.read_csv(path)

### Informasi Umum Data

In [ ]:
df.sample()

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
146975  12/03/2016 23:54  13000       wagon              2008    auto    170   

           Model  Mileage  RegistrationMonth  FuelType          Brand  \
146975  c_klasse   150000                  6  gasoline  mercedes_benz   

       NotRepaired       DateCreated  NumberOfPictures  PostalCode  \
146975          no  12/03/2016 00:00                 0       12101   

                LastSeen  
146975  06/04/2016 13:45

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 354369 entries, 0 to 354368
Data columns (total 16 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   DateCrawled        354369 non-null  object
 1   Price              354369 non-null  int64 
 2   VehicleType        316879 non-null  object
 3   RegistrationYear   354369 non-null  int64 
 4   Gearbox            334536 non-null  object
 5   Power              354369 non-null  int64 
 6   Model              334664 non-null  object
 7   Mileage            354369 non-null  int64 
 8   RegistrationMonth  354369 non-null  int64 
 9   FuelType           321474 non-null  object
 10  Brand              354369 non-null  object
 11  NotRepaired        283215 non-null  object
 12  DateCreated        354369 non-null  object
 13  NumberOfPictures   354369 non-null  int64 
 14  PostalCode         354369 non-null  int64 
 15  LastSeen           354369 non-null  object
dtypes: int64(7), object(

**Dataframe memiliki type yang kurang tepat pada 3 fitur tanggal, tetapi akan di biarkan karna akan kita drop kedepan nya**

In [ ]:
df.describe()

Price  RegistrationYear          Power        Mileage  \
count  354369.000000     354369.000000  354369.000000  354369.000000   
mean     4416.656776       2004.234448     110.094337  128211.172535   
std      4514.158514         90.227958     189.850405   37905.341530   
min         0.000000       1000.000000       0.000000    5000.000000   
25%      1050.000000       1999.000000      69.000000  125000.000000   
50%      2700.000000       2003.000000     105.000000  150000.000000   
75%      6400.000000       2008.000000     143.000000  150000.000000   
max     20000.000000       9999.000000   20000.000000  150000.000000   

       RegistrationMonth  NumberOfPictures     PostalCode  
count      354369.000000          354369.0  354369.000000  
mean            5.714645               0.0   50508.689087  
std             3.726421               0.0   25783.096248  
min             0.000000               0.0    1067.000000  
25%             3.000000               0.0   30165.000000  
50%             6.000000               0.0   49413.000000  
75%             9.000000               0.0   71083.000000  
max            12.000000               0.0   99998.000000

In [ ]:
df.isna().sum()

DateCrawled              0
Price                    0
VehicleType          37490
RegistrationYear         0
Gearbox              19833
Power                    0
Model                19705
Mileage                  0
RegistrationMonth        0
FuelType             32895
Brand                    0
NotRepaired          71154
DateCreated              0
NumberOfPictures         0
PostalCode               0
LastSeen                 0
dtype: int64

**Dataframe memiliki beberapa nilai yang hilang pada fitur dan perlu di cek lebih detail untuk penanganan nya**

In [ ]:
df.duplicated().sum()

262

In [ ]:
df[df.duplicated()].head(1)

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
14266  21/03/2016 19:06   5999       small              2009  manual     80   

      Model  Mileage  RegistrationMonth FuelType       Brand NotRepaired  \
14266  polo   125000                  5   petrol  volkswagen          no   

            DateCreated  NumberOfPictures  PostalCode          LastSeen  
14266  21/03/2016 00:00                 0       65529  05/04/2016 20:47

In [ ]:
df[(df['DateCrawled'] == '21/03/2016 19:06') & (df['Price'] == 5999)]

DateCrawled  Price VehicleType  RegistrationYear Gearbox  Power  \
183    21/03/2016 19:06   5999       small              2009  manual     80   
14266  21/03/2016 19:06   5999       small              2009  manual     80   

      Model  Mileage  RegistrationMonth FuelType       Brand NotRepaired  \
183    polo   125000                  5   petrol  volkswagen          no   
14266  polo   125000                  5   petrol  volkswagen          no   

            DateCreated  NumberOfPictures  PostalCode          LastSeen  
183    21/03/2016 00:00                 0       65529  05/04/2016 20:47  
14266  21/03/2016 00:00                 0       65529  05/04/2016 20:47

**Sepertinya dataframe memiliki data yang duplikat** yang perlu di drop

### Sanity check setiap fitur

#### Fitur VehicleType

In [ ]:
df['VehicleType'].value_counts(dropna=False)

sedan          91457
small          79831
wagon          65166
NaN            37490
bus            28775
convertible    20203
coupe          16163
suv            11996
other           3288
Name: VehicleType, dtype: int64

**Nilai Nan akan dimasukan ke nilai Other**

#### Fitur Gearbox

In [ ]:
df['Gearbox'].value_counts(dropna=False)

manual    268251
auto       66285
NaN        19833
Name: Gearbox, dtype: int64

**Nilai Nan akan dimasukan ke dalam nilai dengan modus terbanyak berdasarkan fitur VehicleType dan Brand**

#### Fitur Power

In [ ]:
df['Power'].describe()

count    354369.000000
mean        110.094337
std         189.850405
min           0.000000
25%          69.000000
50%         105.000000
75%         143.000000
max       20000.000000
Name: Power, dtype: float64

In [ ]:
df[df['Power'] > 500]['Power'].count()

459

**Power dari sebuah mobil diatas 500 sepertinya sangat jarang ditemukan, maka dari itu agar tidak adanya outlier, nilai akan di drop**

#### Fitur Model

In [ ]:
df['Model'].value_counts().count()

250

**Nilai yang unique pada Model sebanyak 250, terlalu banyak untuk dilakukan OHE, oleh karena itu akan dilakukan drop saat pelatihan model LinearRegression, RandomForest dan Decission Tree, tetapi akan tetap digunakan saat melakukan train model dengan LightGBM dan CatBoost**

#### Fitur RegistrationYear

In [ ]:
df[(df['RegistrationYear'] > 2022) | (df['RegistrationYear'] < 1950) ]['RegistrationYear'].count()

351

**sepertinya tidaklah mungkin mendaftarkan kendaraan di masa depan dan di tahun sebelum 1950. oleh karena itu nilai tersebut akan di drop**

#### Fitur Mileage

In [ ]:
df['Mileage'].describe()

count    354369.000000
mean     128211.172535
std       37905.341530
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: Mileage, dtype: float64

**Sepertinya tidak ada masalah pada fitur Mileage**

#### Fitur RegistrationMonth

In [ ]:
df['RegistrationMonth'].value_counts()

0     37352
3     34373
6     31508
4     29270
5     29153
7     27213
10    26099
12    24289
11    24186
9     23813
1     23219
8     22627
2     21267
Name: RegistrationMonth, dtype: int64

**tidak ada masalah pada fitur RegistrationMonth**

#### Fitur FuelType

In [ ]:
df['FuelType'].value_counts(dropna=False)

petrol      216352
gasoline     98720
NaN          32895
lpg           5310
cng            565
hybrid         233
other          204
electric        90
Name: FuelType, dtype: int64

**Nilai yang hilang pada fitur FuelType akan di masukan ke nilai Other**

#### Fitur Brand

In [ ]:
df['Brand'].value_counts(dropna=False).count()

40

**Tidak ada masalah pada Fitur Brand**

#### Fitur NotRepaired

In [ ]:
df['NotRepaired'].value_counts(dropna=False)

no     247161
NaN     71154
yes     36054
Name: NotRepaired, dtype: int64

**Nilai Nan pada fitur NotRepaired akan di isi dengan Unknown**

#### Fitur NumberOfPictures

In [ ]:
df['NumberOfPictures'].value_counts(dropna=False)

0    354369
Name: NumberOfPictures, dtype: int64

**Dikarenakan nilai pada fitur NumberOfPictures hanya ada 0, makan fitur tersebut akan di drop**

In [ ]:
df.columns

Index(['DateCrawled', 'Price', 'VehicleType', 'RegistrationYear', 'Gearbox',
       'Power', 'Model', 'Mileage', 'RegistrationMonth', 'FuelType', 'Brand',
       'NotRepaired', 'DateCreated', 'NumberOfPictures', 'PostalCode',
       'LastSeen'],
      dtype='object')

### Perbaikan Data

#### Drop

In [ ]:
#Drop Duplicates
df.drop_duplicates(inplace = True)

In [ ]:
# Drop Features
drop_col = ['DateCrawled','DateCreated','LastSeen','NumberOfPictures','PostalCode']
df_col_drop = df.drop(columns=drop_col,axis=1)

In [ ]:
df_col_drop.shape

(354107, 11)

In [ ]:
df_col_drop.columns

Index(['Price', 'VehicleType', 'RegistrationYear', 'Gearbox', 'Power', 'Model',
       'Mileage', 'RegistrationMonth', 'FuelType', 'Brand', 'NotRepaired'],
      dtype='object')

#### missing value

In [ ]:
# VehicleType
df_col_drop['VehicleType'].fillna('other',inplace=True)

In [ ]:
#Power
df_col_drop.drop(df_col_drop[df_col_drop['Power'] > 500].index,inplace=True)

In [ ]:
#RegistrationYear
df_col_drop.drop(df_col_drop[
    (df_col_drop['RegistrationYear'] > 2022) |
    (df_col_drop['RegistrationYear'] < 1950)].index,inplace=True)

In [ ]:
#FuelType
df_col_drop['FuelType'].fillna('other',inplace=True)

In [ ]:
#Fungsi memasukan nilai yang hilang pada kolom Gearbox dengan modus terhadap kolom yang diinginkan
def fill_missing (data, column, value_column):
    grouped_values = data.groupby(column)[value_column].agg(lambda x: pd.Series.mode(x)[0]).reset_index()
    size = len(grouped_values)
    for i in range(size):
        group = grouped_values[column][i]
        value = grouped_values[value_column][i]
        data.loc[(data[column]==group) & (data[value_column].isna()),value_column] = value
    return data

In [ ]:
#Gearbox
#Mengisi nilai yang hilang di Gearbox dengan model dan Brand
df_col_drop = fill_missing (df_col_drop,'Model','Gearbox')
df_col_drop = fill_missing (df_col_drop,'Brand','Gearbox')

In [ ]:
df_col_drop['Gearbox'].value_counts(dropna=False)

manual    285189
auto       68122
Name: Gearbox, dtype: int64

In [ ]:
#NotRepaired
df_col_drop['NotRepaired'].fillna('unknown',inplace=True)

In [ ]:
df_col_drop['Model'].fillna('other',inplace=True)

In [ ]:
df_col_drop.isna().sum()

Price                0
VehicleType          0
RegistrationYear     0
Gearbox              0
Power                0
Model                0
Mileage              0
RegistrationMonth    0
FuelType             0
Brand                0
NotRepaired          0
dtype: int64

In [ ]:
#Reset index
df_col_drop.reset_index(drop=True,inplace=True)

#### Pemisahan data

In [ ]:
df_col_drop.shape

(353311, 11)

In [ ]:
# Data untuk Linear Regression, Random Forest, dan TreeClassifier
data_ohe = df_col_drop.drop(columns = 'Model',axis = 1 )
data_ohe = pd.get_dummies(data_ohe)
data_ohe.shape


(353311, 65)

In [ ]:
data_category = df_col_drop.copy()
data_category.shape

(353311, 11)

In [ ]:
category_columns = ['VehicleType','Gearbox','Model','FuelType','Brand','NotRepaired']

data_category[category_columns] = data_category[category_columns].astype('category')

In [ ]:
data_category.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 353311 entries, 0 to 353310
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype   
---  ------             --------------   -----   
 0   Price              353311 non-null  int64   
 1   VehicleType        353311 non-null  category
 2   RegistrationYear   353311 non-null  int64   
 3   Gearbox            353311 non-null  category
 4   Power              353311 non-null  int64   
 5   Model              353311 non-null  category
 6   Mileage            353311 non-null  int64   
 7   RegistrationMonth  353311 non-null  int64   
 8   FuelType           353311 non-null  category
 9   Brand              353311 non-null  category
 10  NotRepaired        353311 non-null  category
dtypes: category(6), int64(5)
memory usage: 15.8 MB


## Pelatihan Model

### Train test split data

#### Data OHE

In [ ]:
#Data OHE
data_ohe_train,data_ohe_valid = train_test_split(data_ohe,test_size=0.3,random_state=12345)
data_ohe_valid,data_ohe_test = train_test_split(data_ohe_valid,test_size=0.5,random_state=12345)
print(data_ohe_train.shape)
print(data_ohe_valid.shape)
print(data_ohe_test.shape)

(247317, 65)
(52997, 65)
(52997, 65)


In [ ]:
features_train =data_ohe_train.drop(['Price'],axis=1)
target_train = data_ohe_train['Price']

features_valid = data_ohe_valid.drop(['Price'],axis=1)
target_valid = data_ohe_valid['Price']

features_test = data_ohe_test.drop(['Price'],axis=1)
target_test = data_ohe_test['Price']

#### Data Category

In [ ]:
data_cat_train,data_cat_valid = train_test_split(data_category,test_size=0.3,random_state=12345)
data_cat_valid,data_cat_test = train_test_split(data_cat_valid,test_size=0.5,random_state=12345)
print(data_cat_train.shape)
print(data_cat_valid.shape)
print(data_cat_test.shape)

(247317, 11)
(52997, 11)
(52997, 11)


In [ ]:
features_train_cat =data_cat_train.drop(['Price'],axis=1)
target_train_cat = data_cat_train['Price']

features_valid_cat = data_cat_valid.drop(['Price'],axis=1)
target_valid_cat = data_cat_valid['Price']

features_test_cat = data_cat_test.drop(['Price'],axis=1)
target_test_cat = data_cat_test['Price']

### Sanity Check

In [ ]:
model = LinearRegression()

In [ ]:
%%time

model.fit(features_train,target_train)

CPU times: user 1.89 s, sys: 476 ms, total: 2.37 s
Wall time: 2.38 s


LinearRegression()

In [ ]:
%%time

pred_train = model.predict(features_train)
pred_valid = model.predict(features_valid)
pred_test = model.predict(features_test)

print('RMSE train =', mean_squared_error(target_train,pred_train)** 0.5)
print('RMSE valid =', mean_squared_error(target_valid,pred_valid)** 0.5)
print('RMSE test =', mean_squared_error(target_test,pred_test)** 0.5)

RMSE train = 2928.2856271857486
RMSE valid = 2914.450624792662
RMSE test = 2910.4835618010575
CPU times: user 144 ms, sys: 253 ms, total: 397 ms
Wall time: 401 ms


### Fungsi check hasil RMSE

In [ ]:
# Data OHE
def hasil_rmse(model):
    pred_train = model.predict(features_train)
    pred_valid = model.predict(features_valid)
    pred_test = model.predict(features_test)

    print('RMSE train =', mean_squared_error(target_train,pred_train)** 0.5)
    print('RMSE valid =', mean_squared_error(target_valid,pred_valid)** 0.5)
    print('RMSE test =', mean_squared_error(target_test,pred_test)** 0.5)

In [ ]:
# Data Category
def hasil_rmse_cat(model):
    pred_train_cat = model.predict(features_train_cat)
    pred_valid_cat = model.predict(features_valid_cat)
    pred_test_cat = model.predict(features_test_cat)

    print('RMSE train =', mean_squared_error(target_train_cat,pred_train_cat)** 0.5)
    print('RMSE valid =', mean_squared_error(target_valid_cat,pred_valid_cat)** 0.5)
    print('RMSE test =', mean_squared_error(target_test_cat,pred_test_cat)** 0.5)

### DecisionTreeRegressor

In [ ]:
modelDT = DecisionTreeRegressor(max_depth=8,random_state=12345)

In [ ]:
%%time

modelDT.fit(features_train,target_train)

CPU times: user 1.24 s, sys: 31.8 ms, total: 1.27 s
Wall time: 1.33 s


DecisionTreeRegressor(max_depth=8, random_state=12345)

In [ ]:
%%time
hasil_rmse(modelDT)

RMSE train = 2201.0670105121762
RMSE valid = 2237.484251037523
RMSE test = 2232.668513482047
CPU times: user 78.9 ms, sys: 32 ms, total: 111 ms
Wall time: 117 ms


### RandomForestRegressor

In [ ]:
modelRF = RandomForestRegressor(n_estimators=20,max_depth=8,max_features='sqrt',
                                random_state=12345)

In [ ]:
%%time

modelRF.fit(features_train,target_train)

CPU times: user 2.82 s, sys: 31.9 ms, total: 2.86 s
Wall time: 2.87 s


RandomForestRegressor(max_depth=8, max_features='sqrt', n_estimators=20,
                      random_state=12345)

In [ ]:
%%time
hasil_rmse(modelRF)

RMSE train = 2557.6226627505785
RMSE valid = 2553.5362455126533
RMSE test = 2558.9724355851713
CPU times: user 628 ms, sys: 27.7 ms, total: 656 ms
Wall time: 666 ms


### LightGBM

In [ ]:
modelLG = lgb.LGBMRegressor(learning_rate=1.5,num_leaves=5,random_state=12345)

#### Data OHE

In [ ]:
%%time

modelLG.fit(features_train,target_train)

CPU times: user 5.09 s, sys: 70 ms, total: 5.16 s
Wall time: 6.28 s


LGBMRegressor(learning_rate=1.5, num_leaves=5, random_state=12345)

In [ ]:
%%time
hasil_rmse(modelLG)

RMSE train = 2001.6100878715665
RMSE valid = 2027.8024285273166
RMSE test = 2038.2137611727894
CPU times: user 1.69 s, sys: 42.9 ms, total: 1.73 s
Wall time: 1.71 s


#### Data Category

In [ ]:
%%time

modelLG.fit(features_train_cat,target_train_cat,categorical_feature='auto')

CPU times: user 3.65 s, sys: 4.64 ms, total: 3.65 s
Wall time: 3.69 s


LGBMRegressor(learning_rate=1.5, num_leaves=5, random_state=12345)

In [ ]:
%%time

pred_train_cat = modelLG.predict(features_train_cat)
pred_valid_cat = modelLG.predict(features_valid_cat)
pred_test_cat = modelLG.predict(features_test_cat)

print('RMSE train =', mean_squared_error(target_train_cat,pred_train_cat)** 0.5)
print('RMSE valid =', mean_squared_error(target_valid_cat,pred_valid_cat)** 0.5)
print('RMSE test =', mean_squared_error(target_test_cat,pred_test_cat)** 0.5)

RMSE train = 1975.7353769090582
RMSE valid = 2024.843288118787
RMSE test = 2022.3274155432102
CPU times: user 1.89 s, sys: 0 ns, total: 1.89 s
Wall time: 1.91 s


### CatBoost

In [ ]:
modelCB =CatBoostRegressor(iterations=25,
                          depth=8,
                          learning_rate=1,
                          loss_function='RMSE',
                          task_type='CPU',
                          verbose=False,
                          random_seed=12345,
                          subsample=0.5
                          )

#### Data OHE

In [ ]:
%%time

modelCB.fit(features_train,target_train)

CPU times: user 1.54 s, sys: 15.7 ms, total: 1.56 s
Wall time: 1.72 s


In [ ]:
%%time
hasil_rmse(modelCB)

RMSE train = 1856.0635252390966
RMSE valid = 1883.6435852710993
RMSE test = 1907.5236078224927
CPU times: user 113 ms, sys: 112 µs, total: 113 ms
Wall time: 111 ms


#### Data Category

In [ ]:
%%time
modelCB.fit(features_train_cat,target_train_cat,cat_features = category_columns)

CPU times: user 4.79 s, sys: 59.6 ms, total: 4.85 s
Wall time: 6.13 s


In [ ]:
%%time
hasil_rmse_cat(modelCB)

RMSE train = 1832.8775431368722
RMSE valid = 1871.3964747620385
RMSE test = 1884.1476499366393
CPU times: user 379 ms, sys: 4.4 ms, total: 383 ms
Wall time: 381 ms


## Analisis Model

### Sanity Check Linear Regression

Dengan model Linear Regression tanpa Tuning Hyperparameter yang di anggap sebagai Sanity check dapat menghasilkan:
1. Waktu training data OHE selama 2.36 s
2. Hasil RMSE yang didapat adalah
    
    RMSE train = 2928.2856271857486 ,
    
    RMSE valid = 2914.450624792662 ,
    
    RMSE test = 2910.4835618010575
3. dengan waktu operasi perhitung RMSE selama 382 ms

### DecisionTreeRegressor

Model DecisionTreeRegressor dengan Tuning Hyperparameter max_depth=8, dapat menghasilkan waktu training selama **1.35 s**, dengan hasil RMSE yang di dapat adalah:
1. RMSE train = 2201.0670105121762
2. RMSE valid = 2237.484251037523
3. RMSE test = 2232.668513482047

yang menghabiskan waktu operasi perhitungan RMSE selama **114 ms**.

Dengan model ini kita mendapatkan **waktu training yang lebih cepat dan hasil RMSE yang lebih rendah**

### RandomForestRegressor

Model RandomForestRegressor dengan Tuning Hyperparameter **n_estimators=20,max_depth=8,max_features='sqrt'**,
dapat menghasilkan waktu training selama **2.74 s**, dengan hasil RMSE yang di dapat adalah:
1. RMSE train = 2557.6226627505785
2. RMSE valid = 2553.5362455126533
3. RMSE test = 2558.9724355851713

yang menghabiskan waktu operasi perhitungan RMSE selama **646 ms**.

Dengan model ini kita mendapatkan **waktu training yang lebih lama dengan DecissionTree dan hasil RMSE yang lebih tinggi**

### LightGBM

Model LightGBM dengan Tuning Hyperparameter **learning_rate=1.5,num_leaves=5,random_state=12345**,
dapat menghasilkan waktu training menggunakan **data OHE** selama **4 s**, dengan hasil RMSE yang di dapat adalah:
1. RMSE train = 2001.6100878715665
2. RMSE valid = 2027.8024285273166
3. RMSE test = 2038.2137611727894

yang menghabiskan waktu operasi perhitungan RMSE selama **1.8 s**.

jika menggunakan **data category** membutuhkan waktu training selama **4.26 s**
1. RMSE train = 1975.7353769090582
2. RMSE valid = 2024.843288118787
3. RMSE test = 2022.3274155432102

yang menghabiskan waktu operasi perhitungan RMSE selama **1.88 s**

Dengan model ini kita mendapatkan **waktu training yang lebih lama dengan DecissionTree tetapi hasil RMSE yang lebih Rendah**

### CatBoost

Model CatBoost dengan Tuning Hyperparameter **iterations=25,depth=8,learning_rate=1,loss_function='RMSE',task_type='CPU',verbose=False,random_seed=12345,subsample=0.5**,
dapat menghasilkan waktu training menggunakan **data OHE** selama **1.64**, dengan hasil RMSE yang di dapat adalah:
1. RMSE train = 1856.0635252390966
2. RMSE valid = 1883.6435852710993
3. RMSE test = 1907.5236078224927

yang menghabiskan waktu operasi perhitungan RMSE selama **101 ms**.

jika menggunakan **data category** membutuhkan waktu training selama **4.85 s**
1. RMSE train = 1832.8775431368722
2. RMSE valid = 1871.3964747620385
3. RMSE test = 1884.1476499366393

yang menghabiskan waktu operasi perhitungan RMSE selama **362 ms**

Dengan model ini kita mendapatkan **waktu training yang sedikit lebih lama dengan DecissionTree tetapi hasil RMSE yang jauh lebih Rendah**

### Kesimpulan

Dari model yang sudah kita coba dengan menggunakan beberapa tipe model dan menggunakan 2 tipe data(OHE & Category).

didapat model yang memiliki kecepatan dan hasil RMSE yang baik adalah **model Catboost dengan data OHE**,

dimana waktu yang dihabiskan untuk training data OHE selama **1.64**, lebih lama sedikit dengan kecepatan waktu training model DecissionTree yang **1.35**.

Tetapi hasil RMSE yang di dapat oleh model Catboost adalah **1832...** yang jauh lebih kecil dibandingkan dengan model DecissionTree yang mendapatkan hasil RMSE sebesar **2201**,